core/base/index_set.cpp

```diff
@@ -0,0 +1,167 @@
+{
+    auto exec = this->get_executor();
+    this->num_stored_indices_ = indices.get_num_elems();
+    exec->run(index_set::make_populate_subsets(
+        this->index_space_size_, &indices, &this->subsets_begin_,
+        &this->subsets_end_, &this->superset_cumulative_indices_));
+}
+
+
+template <typename IndexType>
+bool IndexSet<IndexType>::is_element(const IndexType &input_index) const
+{
+    auto exec = this->get_executor();
+    auto loc_idx =
+        Array<IndexType>(exec, std::initializer_list<IndexType>{input_index});
+    auto glob_idx =
+        Array<IndexType>(exec, std::initializer_list<IndexType>{input_index});
+
+    GKO_ASSERT(this->get_num_subsets() >= 1);
+    exec->run(index_set::make_global_to_local(
+        this->index_space_size_, &this->subsets_begin_, &this->subsets_end_,
+        &this->superset_cumulative_indices_, &glob_idx, &loc_idx));
+    auto is_element_flag = (make_temporary_clone(exec->get_master(), &loc_idx)
+                                .get()
+                                ->get_data()[0]) != -1;
+    return is_element_flag;
+}
+
+
+template <typename IndexType>
+IndexType IndexSet<IndexType>::get_global_index(const IndexType &index) const
+{
+    auto exec = this->get_executor();
+    auto loc_idx =
+        Array<IndexType>(exec, std::initializer_list<IndexType>{index});
+    auto glob_idx =
+        Array<IndexType>(exec, std::initializer_list<IndexType>{index});
+
+    GKO_ASSERT(this->get_num_subsets() >= 1);
+    exec->run(index_set::make_local_to_global(
+        this->index_space_size_, &this->subsets_begin_, &this->subsets_end_,
+        &this->superset_cumulative_indices_, &loc_idx, &glob_idx));
+    auto host_glob_idx = make_temporary_clone(exec->get_master(), &glob_idx)
+                             .get()
+                             ->get_data()[0];
+    return host_glob_idx;
+}
+
+
+template <typename IndexType>
+IndexType IndexSet<IndexType>::get_local_index(const IndexType &index) const
+{
+    auto exec = this->get_executor();
+    auto loc_idx =
+        Array<IndexType>(exec, std::initializer_list<IndexType>{index});
+    auto glob_idx =
+        Array<IndexType>(exec, std::initializer_list<IndexType>{index});
+
+    GKO_ASSERT(this->get_num_subsets() >= 1);
+    exec->run(index_set::make_global_to_local(
+        this->index_space_size_, &this->subsets_begin_, &this->subsets_end_,
+        &this->superset_cumulative_indices_, &glob_idx, &loc_idx));
+    auto host_loc_idx =
+        make_temporary_clone(exec->get_master(), &loc_idx).get()->get_data()[0];
+    return host_loc_idx;
+}
+
+
+template <typename IndexType>
+Array<IndexType> IndexSet<IndexType>::get_global_indices(
+    const Array<IndexType> &local_indices) const
+{
+    auto exec = this->get_executor();
+    auto global_indices =
+        gko::Array<IndexType>(exec, local_indices.get_num_elems());
+
+    GKO_ASSERT(this->get_num_subsets() >= 1);
```

Comment:
Do you need to check if the number of elements of `local_indices` is inferior or equal to `num_stored_indices` or is that not an error?

Reply:
I think if for some reason, someone wants to have the same indices, then it is technically possible that the size of the `local_indices` array is larger than the number of stored indices. So, I would not necessarily check for that.